### PR TITLE
Replace threshold/whitelist keywords by realms file/_class/_package/report

### PR DESCRIPTION
**Pull request by uschi2000**
@jmcampanini @zoubeiri could you take a look at this please?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,11 +22,15 @@ Add the following configuration to `build.gradle`:
...
+    apply plugin: 'java'
     apply plugin: 'jacoco-coverage'
     jacocoCoverage {
-        threshold 0.5  // Enfore minimum code coverage of 50% for every file, class, package, and project.
-        whitelist "MyClass.java"  // Exempt files named MyClass.java from coverage requirements.
+        // Enfore minimum code coverage of 50% for every Java file
+        file 0.5
+
+        // Whitelist files named MyClass.java from coverage requirements.
+        file 0.0, "MyClass.java"
```
Comment (jmcampanini) [inline README.md]: does this mean we're losing the `whitelist` functionality?

File: README.md [hunk 2]
```diff
@@ -35,66 +39,85 @@ project's tests. (Note that `build` depends on `check` which depends on the `jac
...
 #### Configuration
 
-Code coverage requirements can be specified for different scopes:
-- average coverage in a report
-- average coverage in a package
-- coverage in a class
-- coverage in a file
+Code coverage requirements are specified separately for the different "realms" reported by Jacoco:
+- average coverage in a report, specified by the `report` keyword
+- average coverage in a package, specified by the `_package` keyword
```
Comment (jmcampanini) [inline README.md]: having a preceding `_` feels weird. i just read your note below about it, which makes sense. just not sure what's better here.

-`pkg`
-`javaPackage`
-`packageCoverage`

i dunno. not sure if this is worth blocking in, considering there is no super strong alternative

File: README.md [hunk 3]
```diff
@@ -35,66 +39,85 @@ project's tests. (Note that `build` depends on `check` which depends on the `jac
...
+- coverage in a class, specified by the `_class` keyword
+- coverage in a file, specified by the `file` keyword
 
-Scopes are specified either by their exact name as it appears in the Jacoco XML report, for example package scope
-`"org/product/module"`, file scope `"MyClass.java"`, class scope `"org/product/module/MyCLass"`), or by a regular
-expression matching those scope names, for example `~"org.*"` for all packages or class scopes starting with `org`.
+Note that the keywords for package and class realms require a `_` prefix since they are reserved words in Groovy. Each 
+realm contains a number of so-called "scopes": 
+- The file realm contains one scope for every source file, named by file name, e.g., "MyClass.java"
```
Comment (jmcampanini) [inline README.md]: just curious, how does it deal with files with the same name?

File: README.md [hunk 4]
```diff
@@ -35,66 +39,85 @@ project's tests. (Note that `build` depends on `check` which depends on the `jac
...
+- The class realm contains one scope for every (top-level, inner, or anonymous) class, named by fully qualified class
+name, e.g., "org/package/MyClass$1"
+- The package realm contains one scope for every package, named by the package name, e.g., "org/package"
```
Comment (jmcampanini) [inline README.md]: is it `/` delimited or `.` delimited? my guess is that packages are `.` delimited

File: README.md [hunk 5]
```diff
@@ -35,66 +39,85 @@ project's tests. (Note that `build` depends on `check` which depends on the `jac
...
+
+        // Minimum average coverage of 50% in report "my-project; the report name is usually the Gradle project name.
+        report 0.5, "my-project" 
 
         // Scopes can be exempt from all coverage requirements by exact scope name or scope name pattern.
-        whitelist "MyClass.java"
-        whitelist "org/company/module"
-        whitelist ~".*Test.java"
+        file 0.0, "MyClass.java"
+        package 0.0, "org/company/module"
```
Comment (jmcampanini) [inline README.md]: wouldn't this be `_package`?

File: README.md [hunk 6]
```diff
@@ -35,66 +39,85 @@ project's tests. (Note that `build` depends on `check` which depends on the `jac
...
+        file 0.0, ~".*Test.java"
     }
 
-Given the observed code coverage for a scope and coverage type (e.g., _line_ coverage 57% for class
+    // TODO
```
Comment (jmcampanini) [inline README.md]: why is there a floating `TODO` here?

File: README.md [hunk 7]
```diff
@@ -35,66 +39,85 @@ project's tests. (Note that `build` depends on `check` which depends on the `jac
...
-        // Minimum average coverage of 30% in given package.
-        threshold 0.3, "org/company/module"
+        // 10% coverage for all sub-packages of org/package
+        _package 0.1, ~"org/package.*"
+        report 0.8 // 80% coverage required at the report level, i.e., for at average across the project
 
-        // Minimum average coverage of 50% in report "my-project; the report name is usually the Gradle project name.
-        threshold 0.5, "my-project" 
+
+    jacocoCoverage {
```
Comment (jmcampanini) [inline README.md]: i like this example block, but i leave it wondering what the actual spec is for each one.
- `file` uses a `file THRESHOLD, TYPE [, FILE_FILTER]
- `_package` uses `_package THRESHOLD, PACKAGE_NAME`
- etc.

is that anywhere? would that be of value? i don't actually know. i just find it hard to keep straight by looking at just the examples.

File: src/main/groovy/com/palantir/jacoco/CoverageRealm.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.jacoco;
+
+public enum CoverageRealm {
```
Comment (jmcampanini) [inline src/main/groovy/com/palantir/jacoco/CoverageRealm.java:19]: javadoc?

File: src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy [hunk 9]
```diff
@@ -16,6 +16,9 @@
 
 package com.palantir.jacoco
 
+import com.google.common.collect.ArrayListMultimap
+import com.google.common.collect.Multimap
+
 import java.util.regex.Pattern
 
 public class JacocoCoverageExtension {
```
Comment (jmcampanini) [inline src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy]: javadoc?

File: src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy [hunk 10]
```diff
@@ -28,36 +31,72 @@ public class JacocoCoverageExtension {
...
+
+            mc."${realm.realmName}" = { double value, String scope ->
+                threshold(realm, value, scope)
+            }
+
+            mc."${realm.realmName}" = { double value, CoverageType coverageType, String scope ->
+                threshold(realm, value, coverageType, scope)
+            }
+
+            mc."${realm.realmName}" = { double value, Pattern scopePattern ->
```
Comment (jmcampanini) [inline src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy:63]: i love groovy :)

**Comment by jmcampanini** [comment]
done reviewing (sorry, there is no submit button, so i figure i should say when i'm actually done with reading it all :))

File: README.md (hunk 2 above)
Comment (jmcampanini) [inline README.md]: maybe have all prefixed with `_` (fink suggested it!)

File: README.md (hunk 1 above)
Comment (uschi2000) [inline README.md]: yep. syntax is now `file 0.0, "MyClass.java"` or `_class 0.0, "org/package/MyClass"`

File: README.md (hunk 2 above)
Comment (uschi2000) [inline README.md]: done

File: README.md (hunk 3 above)
Comment (uschi2000) [inline README.md]: they will clash; called out more explicitly

File: README.md (hunk 4 above)
Comment (uschi2000) [inline README.md]: /

File: README.md (hunk 5 above)
Comment (uschi2000) [inline README.md]: yep

File: README.md (hunk 6 above)
Comment (uschi2000) [inline README.md]: removed

File: README.md (hunk 7 above)
Comment (uschi2000) [inline README.md]: added syntax definition:

Coverage requirements are written in terms of a realm, a scope specification, and one (or several) of the Jacoco-defined
types of coverage, e.g., lines covered or branches covered. The following syntax variations are supported:

```
<realm> <threshold>
<realm> <threshold>, <coverage type>
<realm> <threshold>, <scope name>
<realm> <threshold>, <coverage type>, <scope name>
<realm> <threshold>, <scope pattern>
<realm> <threshold>, <coverage type>, <scope pattern>
```

, where:
- `<realm>` is one of `_file`, `_class`, `_package`, `_report`
- `<threshold>` is a Groovy `double`
- `<coverage type>` is one of `BRANCH`, `CLASS`, `COMPLEXITY`, `INSTRUCTION`, `LINE`, `METHOD`
- `<scope name>` is a Groovy string
- `<scope pattern>` is a Groovy regex pattern

File: src/main/groovy/com/palantir/jacoco/CoverageRealm.java (hunk 8 above)
Comment (uschi2000) [inline src/main/groovy/com/palantir/jacoco/CoverageRealm.java:19]: done

File: src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy (hunk 9 above)
Comment (uschi2000) [inline src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy]: done

File: src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy (hunk 10 above)
Comment (uschi2000) [inline src/main/groovy/com/palantir/jacoco/JacocoCoverageExtension.groovy:63]: love-hate?

**Pull request merged**